### DISCUSSION TITLE
/hyndsight/forecast4/

### DISCUSSION THREAD

**Discussion by robjhyndman** [robjhyndman/robjhyndman.com#129] [Announcements]
New in forecast 4.0 
 https://robjhyndman.com/hyndsight/forecast4/

**Comment by robjhyndman** [comment]
```header
Originally posted by fernando on 2012-12-03 17:18:57
````
It would be nice to have an automatic cross-validation function, given
the fitted model and the validation parameters (training-set size, max
Horizon, etc). Anyway, great job in the new version!

**Reply by robjhyndman** [reply]
```header
Originally posted by Zachary Deane-Mayer on 2012-12-03 21:08:17
````
There\'s some code here you can try out:
[http://robjhyndman.com/rese\...](http://robjhyndman.com/researchtips/tscvexample/)

You can also take a look at the R package I am developing to do this:
[https://github.com/zachmaye\...](https://github.com/zachmayer/cv.ts)

**Comment by robjhyndman** [comment]
```header
Originally posted by Samik on 2012-12-03 17:37:02
````
Hi Rob, thanks for the notification. Indeed, the \'What\'s new\' section
rarely gets read. Anways, I was wondering if you have an updated version
of the original paper which detailed the auto.arima() algorithm, that
includes the improvements and changes?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2012-12-03 22:09:18
````
The description at
[http://otexts.com/fpp/8/](http://otexts.com/fpp/8/)
is up-to-date.

**Comment by robjhyndman** [comment]
```header
Originally posted by Leandro Callegari Coelho on 2012-12-05 02:15:08
````
Rob, congrats for the development. I like this package a lot. By the
way, whenever a new version is available, the first thing I look for is
the changelog/what\'s new section!

**Comment by robjhyndman** [comment]
```header
Originally posted by Shea Parkes on 2012-12-11 16:09:07
````
Thanks for all your hard work Rob! I appreciate the subtler default
colors, but you\'ll need to go back and update your online book to stop
mentioning yellow/orange now:
[http://otexts.com/fpp/1/4/](http://otexts.com/fpp/1/4/)

**Comment by robjhyndman** [comment]
```header
Originally posted by Jerome on 2012-12-18 15:38:40
````
Thank you Rob for the function dshw. When I get the point of forecast,
how could I get the forecast interval (low 0.80 and upper 0.80)? Thank
you

**Reply by robjhyndman** [reply]
```header
Originally posted on 2012-12-19 00:42:58
````
I have not implemented prediction intervals for dshw(). Use bats() to
get PI for an equivalent model.

**Comment by robjhyndman** [comment]
```header
Originally posted by BMiner on 2012-12-23 19:05:47
````
Hey Rob, You mention prediction intervals with ANN via simulation. Do
you have any sources to reference?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2012-12-23 22:37:21
````
No, but a neural network is essentially a nonlinear autoregression and
simulation is the most common way to get PI for other nonlinear
autoregression models.

**Reply by robjhyndman** [reply]
```header
Originally posted by Bminer on 2012-12-24 02:59:47
````
Do you think bootstrapping would be a viable method (using the boot
package and tsboot)? I am thinking of bootstrapping the series and then
using each of the n models created to predict a new case - and thus
getting a prediction distribution.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2012-12-24 03:03:05
````
No. The bootstrapped series won\'t allow prediction conditional on the
last few values of the observed series.

**Comment by robjhyndman** [comment]
```header
Originally posted by Joseph Johnson on 2013-01-16 19:14:52
````
Dr. Hyndman,

Does the ets function in forecast package v4.0 constrain the parameters
to be;\
0\&lt;α\&lt;1, 0\&lt;β\&lt;α, 0\&lt;γ\&lt;1-α ?\
Would you recommend this constraint for all models?\
Does the package go even further to constraint the certain models as
mentioned in Chapter 10 of your exponential smoothing book?

Thank you for the package. It has been very helpful to me!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-01-16 20:30:10
````
Try reading the help file. It is all explained. The constraints you
mention are implemented using bounds=&quot;usual&quot;. The constraints in ch10
of our book are implemented using bounds=&quot;admissible&quot;. The default is
bounds=&quot;both&quot; meaning both sets of constraints are applied.

**Reply by robjhyndman** [reply]
```header
Originally posted by Joseph Johnson on 2013-01-16 23:55:32
````
Thank you.

**Comment by robjhyndman** [comment]
```header
Originally posted by Francesco Gianferrari Pini on 2013-01-20 20:32:40
````
Hi! How can I extract the time series for the seasonal component(s) on a
TBATS forecast? I can see them in the plot!

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-01-20 20:47:48
````
Look at plot.tbats() and you will see how they are constructed.

**Comment by robjhyndman** [comment]
```header
Originally posted by Jerome on 2013-02-12 23:58:01
````
Bonjour Rob! I am running a bats function and I get this error:

Error in checkForRemoteErrors(val) : 6 nodes produced errors; first
error: argument is of length zero\
I have no clue to solve it. Thank you in advance for your tips.\
Jerome

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-02-13 00:09:51
````
Try using debug() to track where the error is occurring.

**Comment by robjhyndman** [comment]
```header
Originally posted by SimonMN on 2013-05-16 11:53:45
````
Hi Prof Hyndman,

Is there any way to add exogenous inputs to the nnetar function in the
same way you can add exogenous variables to the Arima or auto.arima
functions?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-05-16 11:58:09
````
No, but it isn\'t hard to do it using the avNNet function from the caret
package. Basically nnetar() is forecasting using an average of 20 neural
nets where the inputs are lagged values of the time series. You would
just need to add the xreg variables as inputs as well.

**Reply by robjhyndman** [reply]
```header
Originally posted by llr on 2014-05-12 13:57:35
````
Hi Prof Hyndman,\
Can you point to an example of this technique? In your textbook, you
show an example of avNNet() on a cross-sectional dataset, which is
helpful. To implement avNNet() on a time series dataset, would you
create a dataframe that holds not only dummy variables for
holiday/seasonal effects but also the lagged variables? Eg something
like avNNet(point_forecast \~ lagged_1 + lagged_2 + lagged_3 +
holiday_1) where lagged_1,..,lagged_3 are the lagged variables for that
daily forecast? (And thus not use nnetar() at all?)

**Reply by robjhyndman** [reply]
```header
Originally posted on 2014-05-12 21:38:03
````
Yes

**Comment by robjhyndman** [comment]
```header
Originally posted by Christian Miller on 2013-05-17 03:09:48
````
I don\'t have anything substantive to add, but I want to give a huge
thanks to you Rob. This has been a helpful and useful package in
learning about forecasting and implementing ideas. And you continue to
add functionality faster than I can keep up: it\'s a great problem to
have.

Greatly appreciated (and the book too!)

**Comment by robjhyndman** [comment]
```header
Originally posted by Tuan on 2013-07-17 17:58:32
````
Hi Rob,

is there any roadmap for computing the prediction intervals of NNETAR
models? For some of our data (machine and network activity monitoring),
the NNETAR models seem to render the best result.

Thank you so much for your great work.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-07-17 20:59:51
````
I wasn\'t planning on adding them any time soon. It requires simulation
of future sample paths. If someone was to write the code, I would add it
to the package!

**Reply by robjhyndman** [reply]
```header
Originally posted by Tuan on 2013-07-23 17:56:48
````
Thank you for much for the prompt response. I have reviewed several
approaches to to calculate the prediction-intervals and seems to like
the approach based on Gheyas and Smith\'s papers the most.
([http://pdf.aminer.org/000/2\...](http://pdf.aminer.org/000/262/006/forecasting_time_series_combining_machine_learning_and_box_jenkins_time.pdf)
and
[http://www.cs.stir.ac.uk/\~l\...](http://www.cs.stir.ac.uk/~lss/recentpapers/NEUCOM_12341.PDF) &quot;http://www.cs.stir.ac.uk/~lss/recentpapers/NEUCOM_12341.PDF)&quot;).
Basically, they use two NN, the first one is for forecasting conditional
mean (point forecast), the second one is for forecasting conditional
variance. The input of the second NN is the residual series. The
prediction intervals can be calculated based on the conditional mean and
conditional variance. What do you think about Gheyas and Smith\'s
approach?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-07-23 23:14:01
````
I\'m unfamiliar with those papers, but any approach based only on means
and variances will need to assume that the forecast distributions are
normally distributed which is unlikely. A simulation approach does not
make that assumption.

**Comment by robjhyndman** [comment]
```header
Originally posted by Peter  on 2013-08-04 10:35:57
````
There seems to be a bug in the bats() function.

\&gt;fit\&lt;-bats(mydata,use.box.cox=FALSE,use.trend=TRUE,use.damped.trend=FALSE,seasonal.periods=c(24,168))

This command gives a warning:

optim() did not converge.

The command works without use.damped.trend=FALSE.

regards\
Peter

**Reply by robjhyndman** [reply]
```header
Originally posted by Peter on 2013-08-04 10:40:49
````
I am trying to use bats() in place of dshw().

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-08-04 11:52:28
````
What makes you think it is a bug?

**Reply by robjhyndman** [reply]
```header
Originally posted by Peter on 2013-08-04 12:25:24
````
Because it should work without the damped trend. That\'s my opinion. I
may be wrong.

**Reply by robjhyndman** [reply]
```header
Originally posted by Peter on 2013-08-04 14:10:03
````
dshw() produces forecasts without warnings for the same data.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-08-04 22:37:36
````
It does work with the damped trend. It\'s only a warning, and it only
says that the optimization has not yet converged.

**Comment by robjhyndman** [comment]
```header
Originally posted by Leo on 2013-08-08 13:40:47
````
Hi,\
Congratulations on creating a wonderful forecast package. I have
electricity demand data containing two seasonalities. When I do
tbats(elec), I get just one seasonal period parameters. If I give both
seasonal periods as input then I get parameters for both. But in the
later case, I am not using full power of the tbats model. Can you please
say something in this regard.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-08-08 13:45:17
````
I don\'t know what you mean by not using the full power of the tbats
model.

**Reply by robjhyndman** [reply]
```header
Originally posted by Leo on 2013-08-08 14:07:08
````
I mean let tbats decide using the criteria mentioned in the article.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-08-08 14:08:50
````
tbats doesn\'t decide on the seasonal periods. It chooses the number of
Fourier terms to use for each of the seasonal periods.

**Reply by robjhyndman** [reply]
```header
Originally posted by Leo on 2013-08-08 14:13:09
````
Then how can it find the complex seasonality not visible in the plot. I
remember something similar mentioned in the article.

**Reply by robjhyndman** [reply]
```header
Originally posted by Peter on 2013-08-09 02:31:03
````
You have an option of entering complex seasonal periods. TBATS is not
going to find them for you. TBATS only finds the parameters. There are
no other functions in R that allows this.

**Comment by robjhyndman** [comment]
```header
Originally posted by bayes40 on 2013-08-18 04:41:20
````
Hi Rob:\
Does the forecast package you created handle rational distributed lag
models? I am developing hourly transfer functions that attempt to
predict demand for every hour of the day using 1 1/2 years of daily
demand data with 1 1/2 years of input advertising impression data. I am
also controlling for seasonal variables including weekly base level
demand patterns. I am assuming ARMA errors for the error terms, and I
believe that a Koyck model is all I need to capture the influence of
past advertising on demand (request for quotes). I want to use
Auto.Arima, but not sure I am supposed to use it with for distributive
lag specifications.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-08-18 11:08:07
````
No, although it does handle lagged covariates via the xreg argument in
the Arima() and auto.arima() function. It does not handle a Koyck model,
or any other model with rational lag structure for a covariate. The TSA
package will fit the model, but not produce forecasts. I hope to
implement these models some time, but haven\'t done so yet.

**Reply by robjhyndman** [reply]
```header
Originally posted by bayes40 on 2013-08-19 00:00:28
````
Thank you so much for the quick response. Love your text book

**Reply by robjhyndman** [reply]
```header
Originally posted by bayes40 on 2013-08-19 00:17:15
````
Last question: if you were building a model of advertisings effect on
demand could you build an valid model using your auto arima and forecast
package by adding lagged predictors or is there some added value in
using the TSA packages?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-08-19 01:59:39
````
You can use auto.arima. See
[http://otexts.com/fpp/9/1/](http://otexts.com/fpp/9/1/)
Example 9.3.

**Comment by robjhyndman** [comment]
```header
Originally posted by Tuan on 2013-08-26 17:05:43
````
Hi Rob,

we are using your package \'forecast\' to forecast millions of time
series on hourly basis and would like to explore the option to use the
incremental update of of the tbats models to speed up the processing
(i.e. given a tbats model at the time point Tn, we would like to update
the model when the data point for Tn+1 arrives). I understand that this
feature is currently not available, and we will have to look at the
source code and modify it. Before we begin, can you please advise if
it\'s feasible? If yes, please give us some hints where to modify the
logic. Definitely, it\'s working for us, we would love to contribute
back to the package.

Thanks a lot for your time

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-08-26 22:27:10
````
If you want to re-estimate the parameters, then it is not really
feasible as the parameter optimization is too slow. But if you want to
keep the current parameters, but compute the new forecasts on the basis
of the new data, then it should be possible. That is on my list of
proposed features to add to the package, but it probably won\'t happen
in the near future. You will need to skip the parameter optimization and
model selection part of tbats, but still compute fitted values and
residuals.

**Reply by robjhyndman** [reply]
```header
Originally posted by Tuan on 2013-08-27 16:16:29
````
Thank you so much Rob for the response.

**Comment by robjhyndman** [comment]
```header
Originally posted by Padarn on 2013-09-10 03:51:25
````
Hi Rob. I was going to raise this on github, but it seemed to minor for
an \'issue\'.

Is there any reason not to have a way of specifying automatic selection
of the box-cox transformation parameter (and potentially also a shift)?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-09-10 03:55:56
````
I\'ve not come across an automatic method for the Box-Cox transformation
that is reliable enough to make it the default. You can use
BoxCox.lambda() which is not bad, but I get better forecasts on average
without it.

**Comment by robjhyndman** [comment]
```header
Originally posted by mohammed jarad on 2013-09-14 11:56:43
````
Hello\
My name is mohammed jarad,I\'m student in applied statistical ,\
I have do my research about time series ,neural networks.\
I have problem to use code of R about this code (nnetar)i don\'t know
haw it is\
use .\
can you help me \&gt;\
to forecast using this code(nnetar) .thinks you

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-09-15 22:38:20
````
Read the help file.

**Reply by robjhyndman** [reply]
```header
Originally posted by mohammed jarad on 2013-09-16 05:20:19
````
Thank you..

**Reply by robjhyndman** [reply]
```header
Originally posted by mohammed jarad on 2013-09-16 06:12:59
````
Hi Rob, I\'m do my code for my data .\
ff\&lt;- nnetar(zz, P=1, size=10, repeats=20, lambda=0.05)

but i have problem that i have not find any AIC or BIC .\
I like to tell you what i do?\
I find to (comparison between Box-Jenkins methodology and Neural
Networks to predict) I\'m finish to predict by Box-Jenkins but I have
problem to find predict by Neural Networks ..

can you help my to comparison between them to my data..\
thank you so much Rob.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-09-16 06:14:01
````
Neural nets are not based on likelihood estimation, so there cannot be
an AIC or BIC. Use a rolling hold out sample for comparisons instead.

**Reply by robjhyndman** [reply]
```header
Originally posted by mohammed jarad on 2013-09-16 06:43:03
````
Thank you Rob,\
I\'m going to do it.

**Comment by robjhyndman** [comment]
```header
Originally posted by Matt Weller on 2013-09-15 19:37:31
````
I\'m finding an error when passing an lm model to the accuracy function.
In R I can\'t seem to access (to step through) the functions
forecasterrors(f, x, test) and getResponse(f). Is this code available
and where would I be able to find it? If this debugging doesn\'t solve
it I will prepare a question for stackoverflow. The error I\'m getting
kicks in at getResponse(f) as follows:

Error in \`\[.data.frame\`(model.frame(object\$model), , responsevar) :
undefined columns selected

I can see that the lm formula and data.table contain only the relevant
columns correctly named and I need to dig further into the function
which is defined in R as:

function (object, \...)\
UseMethod(&quot;getResponse&quot;)\
\&lt;bytecode: 0x0c0f93f4=&quot;&quot;\&gt;\
\&lt;environment: namespace:forecast=&quot;&quot;\&gt;

Many thanks for the forecast package, it is awesome and I\'ve only
scratched the surface at this point.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-09-15 22:38:02
````
Can you please indicate what version of the package you are using, and
provide a minimal example so I can replicate the problem.

**Comment by robjhyndman** [comment]
```header
Originally posted by mohammed jarad on 2013-09-19 04:44:09
````
Hi Rob, Can you please sent me any think help me on my research\
about (nnetar) I don't find any think to help my please Rob .I'm going
to find\
forecasting by (nnetar) to comparative neural networks with Box-Jenkens

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-09-19 04:54:18
````
See
[http://otexts.com/fpp/9/3/](http://otexts.com/fpp/9/3/)

**Reply by robjhyndman** [reply]
```header
Originally posted by mohammed jarad on 2013-09-21 12:30:33
````
Thank you Rob.

**Reply by robjhyndman** [reply]
```header
Originally posted by mohammed jarad on 2013-11-16 08:16:47
````
Hello Prof. Hyndman, i do my code. I\'m generating an Arima model in
time series. Let\'s suppose I have n observations. The predict.Arima
function from forecast package just make predictions for n + 1
observation on.

I need to make a prediction for the n value

(I want to predict the real data)

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-11-17 00:29:08
````
fitted() will give one-step forecasts of the data used to estimate the
model.

**Comment by robjhyndman** [comment]
```header
Originally posted by Isabella Ghement on 2013-12-12 19:27:20
````
Dear Rob,

When one uses bootstrapping to derive forecast intervals, the
forecast.Arima() function in the forecast package seems to report
prediction intervals computed as follows:

qq \&lt;- qnorm(0.5 \* (1 + level\[i\]/100))\
lower\[, i\] \&lt;- pred\$pred - qq \* pred\$se\
upper\[, i\] \&lt;- pred\$pred + qq \* pred\$se

In the above, pred\$pred and pred\$se represent the mean and standard
deviation of the point forecasts derived via bootstrapping.

Is there any reason why the normal distribution quantiles are used in
the derivation of these forecast intervals instead of the quantiles
corresponding to the bootstrap distribution of the point forecasts?

Also, if the bootstrap distribution of the point forecasts is non-normal
(e.g., multi-modal), the standard deviation of this distribution may not
be a sensible measure of spread. Similarly, the mean of this
distribution may not be a sensible measure of centre.

Thank you very much for any clarifications you can provide.\
Isabella

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-12-16 10:46:50
````
Yes, the quantiles would be better. I\'ll add it to the list of things
to do.

**Reply by robjhyndman** [reply]
```header
Originally posted by Isabella Ghement on 2013-12-16 19:25:39
````
Thanks so much, Rob! This sounds reasonable. It would also be helpful to
the users to have access to the bootstrap values of the point forecasts,
so they can examine their distribution and assess how close to or far
from normality this distribution actually is.

The other thing I was wondering is as follows:

If the bootstrap distribution of the point forecasts looks really
&quot;wild&quot; (e.g., several peaks, some gaps, asymmetry, etc.), does it
still make sense to report the default point forecast produced by
forecast.Arima() (which is in essence the &quot;mean&quot; value of the
bootstrap distribution of the point forecasts)? Or perhaps it would be
better to report instead the &quot;median&quot; value of the bootstrap
distribution of the point forecasts, along with the quantile-based
forecast interval?

**Reply by robjhyndman** [reply]
```header
Originally posted on 2013-12-16 22:16:48
````
You can already get the bootstrap values via the simulate.Arima
function.

If the distribution of future values looks wild, then you should report
the whole distribution, not just a point estimate or an interval
estimate.

**Comment by robjhyndman** [comment]
```header
Originally posted by iris1873 on 2014-07-17 19:16:58
````
It seems that the nnetar forecasts differ a lot with different sample
size of data we use to build model. If the goal is to improve forecast
accuracy, is there a way to determin the best sample size to build
model? Thanks!

**Comment by robjhyndman** [comment]
```header
Originally posted by Camelo on 2017-08-17 02:24:40
````
Hello Rob,

I\'m having trouble with the nnetar function, by the way. The fact is
that you use a time series of 1 year with data every 10 min, and already
does a transformation for xts, but when using nnetar (timeseries) the
error appears: Error in nnet.default (x, y, linout = Linout, trace =
trace, \...):\
Many (1201) pesos

can you help me? thank you

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-08-17 03:54:21
````
Please provide a minimal reproducible example and report at
[https://github.com/robjhynd\...](https://github.com/robjhyndman/forecast/issues)

**Comment by robjhyndman** [comment]
```header
Originally posted by Kalpana on 2017-08-17 05:56:29
````
hello Rob,\
Is there any way to capture the optimal values of P,p and k in the
NNAR(p,P,k) model.

**Reply by robjhyndman** [reply]
```header
Originally posted on 2017-08-17 09:39:31
````
Look at what is stored using ls(object). You will see p, P and size are
elements.
